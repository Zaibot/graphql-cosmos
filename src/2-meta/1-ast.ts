import {
  DefinitionNode,
  DocumentNode,
  FieldDefinitionNode,
  ListTypeNode,
  NamedTypeNode,
  ObjectTypeDefinitionNode,
  ObjectTypeExtensionNode,
  TypeNode,
} from 'graphql'
import { isListTypeNode } from 'graphql-tools'
import { Directives } from '../1-graphql/2-ast'
import { isWhereOp, WhereOps } from '../6-datasource/x-where'

export interface GraphQLCosmosSchemaFromAst {
  types: GraphQLCosmosTypeFromAst[]
}

export interface GraphQLCosmosTypeFromAst {
  external: boolean
  typename: string
  database: string | null
  container: string | null
  fields: GraphQLCosmosFieldFromAst[]
}

export interface GraphQLCosmosFieldFromAst {
  external: boolean
  fieldname: string
  returnTypename: string
  returnMany: boolean
  container: string | null
  database: string | null
  ours: string | null
  theirs: string | null
  whereOps: WhereOps[] | null
  whereOurs: string | null
  sortable: boolean | null
  sortOurs: string | null
  pagination: string | null
}

export function getGraphQLCosmosSchemaFromGraphQL(objectType: DocumentNode): GraphQLCosmosSchemaFromAst {
  const internalTypes = (objectType.definitions ?? []).filter(isObjectTypeDefinition).map(getGraphQLCosmosTypeFromGraphQL)
  const externalTypes = (objectType.definitions ?? []).filter(isObjectTypeExtension).map(getGraphQLCosmosTypeFromGraphQL)

  return {
    types: [...internalTypes, ...externalTypes],
  }
}

export function getGraphQLCosmosTypeFromGraphQL(objectType: ObjectTypeDefinitionNode | ObjectTypeExtensionNode): GraphQLCosmosTypeFromAst {
  const typename = objectType.name.value
  const fields = (objectType.fields ?? []).map(getGraphQLCosmosFieldFromGraphQL)
  const database = Directives.cosmosDatabaseDirective(objectType.directives ?? []) ?? null
  const container = Directives.cosmosContainerDirective(objectType.directives ?? []) ?? null

  return {
    external: objectType.kind === `ObjectTypeExtension`,
    typename: typename,
    database,
    container,
    fields,
  }
}

export function getGraphQLCosmosFieldFromGraphQL(field: FieldDefinitionNode): GraphQLCosmosFieldFromAst {
  const returnTypename = astGetNamedType(field.type).name.value
  const returnMany = isListTypeNode(astGetNullable(field.type))
  const database = Directives.cosmosDatabaseDirective(field.directives ?? []) ?? null
  const container = Directives.cosmosContainerDirective(field.directives ?? []) ?? null
  const pagination = Directives.cosmosPaginationDirective(field.directives ?? []) ?? null
  const fieldname = field.name.value
  const ours = Directives.cosmosOursDirective(field.directives ?? []) ?? null
  const theirs = Directives.cosmosTheirsDirective(field.directives ?? []) ?? null
  const whereOps = Directives.whereOpDirective(field.directives ?? []).filter(isWhereOp) ?? null
  const whereOurs = Directives.whereOursDirective(field.directives ?? []) ?? null
  const sortable = Directives.sortDirective(field.directives ?? []) ?? null
  const sortOurs = Directives.sortOursDirective(field.directives ?? []) ?? null

  return {
    external: false,
    returnTypename,
    returnMany,
    database,
    container,
    fieldname,
    ours,
    theirs,
    whereOps,
    whereOurs,
    sortable,
    sortOurs,
    pagination,
  }
}

function astGetNamedType(type: TypeNode): NamedTypeNode {
  while (type.kind === `ListType` || type.kind === `NonNullType`) {
    type = type.type
  }
  return type
}

function isObjectTypeDefinition(n: DefinitionNode): n is ObjectTypeDefinitionNode {
  return n.kind === `ObjectTypeDefinition`
}

function isObjectTypeExtension(n: DefinitionNode): n is ObjectTypeExtensionNode {
  return n.kind === `ObjectTypeExtension`
}

function astGetNullable(type: TypeNode): ListTypeNode | NamedTypeNode {
  while (type.kind === `NonNullType`) {
    type = type.type
  }
  return type
}
