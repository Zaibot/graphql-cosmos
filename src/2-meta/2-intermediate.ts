import { WhereOps } from '../6-datasource/x-where'
import { GraphQLCosmosInitError } from '../x-error/init-error'
import { GraphQLCosmosFieldFromAst, GraphQLCosmosSchemaFromAst, GraphQLCosmosTypeFromAst } from './1-ast'

export interface MetaSchema {
  types: MetaType[]
}

export interface MetaType {
  external: boolean
  typename: string
  database: string | null
  container: string | null
  fields: MetaField[]
}

export interface MetaField {
  fieldname: string
  returnMany: boolean
  returnTypename: string
  container: string | null
  database: string | null
  ours: string | null
  theirs: string | null
  pagination: boolean | null
  whereOps: WhereOps[] | null
  whereOurs: string | null
  sortable: boolean | null
  sortOurs: string | null
  kind: MetaFieldKind
}

export type MetaFieldKind = 'many-root' | 'one-ours' | 'many-ours' | 'one-theirs' | 'many-theirs' | 'embedded'

export function getMetaSchema(objectType: GraphQLCosmosSchemaFromAst): MetaSchema {
  try {
    const types = objectType.types.map(getMetaType).sort((a, b) => a.typename.localeCompare(b.typename))

    return {
      types,
    }
  } catch (ex) {
    throw new GraphQLCosmosInitError(`Error while generating meta schema`, ex)
  }
}

export function getMetaType(type: GraphQLCosmosTypeFromAst): MetaType {
  try {
    return {
      external: type.external,
      typename: type.typename,
      database: type.database,
      container: type.container,
      fields: (type.fields ?? []).map(getMetaField).sort((a, b) => a.fieldname.localeCompare(b.fieldname)),
    }
  } catch (ex) {
    throw new GraphQLCosmosInitError(`Type: ${type.typename}`, ex)
  }
}

export function getMetaField(field: GraphQLCosmosFieldFromAst): MetaField {
  try {
    return {
      returnTypename: field.returnTypename,
      fieldname: field.fieldname,
      database: field.database,
      pagination:
        field.pagination === 'on'
          ? true
          : field.pagination === 'off'
          ? false
          : field.returnMany && field.container !== null && field.ours === null,
      container: field.container,
      ours: field.ours,
      theirs: field.theirs,
      whereOps: field.whereOps,
      whereOurs: field.whereOurs,
      sortable: field.sortable,
      sortOurs: field.sortOurs,
      returnMany: field.returnMany,
      kind: toRelationKind(!!field.container, !!field.ours, !!field.theirs, field.returnMany),
    }
  } catch (ex) {
    throw new GraphQLCosmosInitError(`Field: ${field.fieldname}`, ex)
  }
}

function toRelationKind(container: boolean, ours: boolean, theirs: boolean, list: boolean): MetaFieldKind {
  if (ours && theirs) {
    throw Error(`expects ours or theirs not both`)
  } else if (ours && list) {
    // if (!container) throw Error(`expects container for many-ours`)
    return 'many-ours'
  } else if (theirs && list) {
    // if (!container) throw Error(`expects container for many-theirs`)
    return 'many-theirs'
  } else if (ours) {
    // if (!container) throw Error(`expects container for one-ours`)
    return 'one-ours'
  } else if (theirs) {
    // if (!container) throw Error(`expects container for one-theirs`)
    return 'one-theirs'
  } else if (list && container) {
    return 'many-root'
  } else {
    if (container) throw Error(`expects no container for embedded`)
    if (ours) throw Error(`expects no ours for embedded`)
    if (theirs) throw Error(`expects no theirs for embedded`)
    return 'embedded'
  }
}
