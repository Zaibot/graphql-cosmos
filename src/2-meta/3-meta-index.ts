import { fail } from '../typescript'
import { MetaField, MetaSchema, MetaType } from './2-intermediate'

export class MetaIndex {
  readonly typeMap: ReadonlyMap<string, MetaType>
  readonly fieldMap: ReadonlyMap<string, MetaField>
  readonly sortFieldMap: ReadonlyMap<string, MetaField>
  readonly whereFieldMap: ReadonlyMap<string, MetaField>

  constructor(schema: MetaSchema) {
    this.typeMap = new Map(schema.types.map((t) => [t.typename, t]))
    this.fieldMap = new Map(schema.types.flatMap((t) => t.fields.map((f) => [`${t.typename}.${f.fieldname}`, f])))
    this.sortFieldMap = this.buildSortMap(schema)
    this.whereFieldMap = this.buildWhereMap(schema)
  }

  private buildWhereMap(schema: MetaSchema) {
    const whereFieldMap = new Map()
    for (const type of schema.types) {
      for (const field of type.fields) {
        whereFieldMap.set(`${type.typename}.${field.fieldname}`, field)
        whereFieldMap.set(`${type.typename}.${field.ours ?? field.fieldname}`, field)
        whereFieldMap.set(`${type.typename}.${field.whereOurs ?? field.ours ?? field.fieldname}`, field)
      }
    }
    return whereFieldMap
  }

  private buildSortMap(schema: MetaSchema) {
    const sortFieldMap = new Map<string, MetaField>()
    for (const type of schema.types) {
      for (const field of type.fields) {
        sortFieldMap.set(`${type.typename}.${field.fieldname}`, field)
        sortFieldMap.set(`${type.typename}.${field.ours ?? field.fieldname}`, field)
        sortFieldMap.set(`${type.typename}.${field.sortOurs ?? field.ours ?? field.fieldname}`, field)
      }
    }
    return sortFieldMap
  }

  get allTypes() {
    return Array.from(this.typeMap.values())
  }

  get whereableTypes() {
    return Array.from(this.typeMap.values()).filter((t) => t.fields.some((x) => x.whereOps?.length))
  }

  get sortableTypes() {
    return Array.from(this.typeMap.values()).filter((t) => t.fields.some((x) => x.sortable))
  }

  get pageableTypes() {
    return unique(
      Array.from(this.typeMap.values())
        .flatMap((x) => x.fields)
        .filter(
          (x) =>
            (x.pagination && this.type(x.returnTypename)) ||
            (x.container === null && this.type(x.returnTypename)?.container)
        )
        .map((x) => this.requireType(x.returnTypename))
    )
  }

  type(typename: string) {
    return this.typeMap.get(typename)
  }

  field(typename: string, fieldname: string) {
    return this.fieldMap.get(`${typename}.${fieldname}`)
  }

  whereField(typename: string, fieldname: string) {
    return this.whereFieldMap.get(`${typename}.${fieldname}`)
  }

  sortField(typename: string, fieldname: string) {
    return this.sortFieldMap.get(`${typename}.${fieldname}`)
  }

  requireType(typename: string) {
    return this.typeMap.get(typename) ?? fail(`meta type missing: ${typename}`)
  }

  requireField(typename: string, fieldname: string) {
    return this.fieldMap.get(`${typename}.${fieldname}`) ?? fail(`meta field missing: ${typename}.${fieldname}`)
  }
}

function unique<T>(array: T[]) {
  return Array.from(new Set(array).values())
}
