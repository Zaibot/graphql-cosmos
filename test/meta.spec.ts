import gql from 'graphql-tag'
import { getGraphQLCosmosSchemaFromGraphQL } from '../src/2-meta/1-ast'
import { getMetaSchema } from '../src/2-meta/2-intermediate'

const dummyTypeDefs = gql`
  type Query {
    dummies: [Dummy] @cosmos
  }

  type Dummy @cosmos(database: "Test", container: "Dummies") {
    id: ID! @where(op: "eq in")
    text: String
  }
`

describe(`@cosmos`, () => {
  it(`should be retrieve all items`, async () => {
    const meta = getMetaSchema(getGraphQLCosmosSchemaFromGraphQL(dummyTypeDefs))

    expect(meta).toMatchSnapshot()
  })
})
