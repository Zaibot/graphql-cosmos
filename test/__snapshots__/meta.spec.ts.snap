// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@cosmos should be retrieve all items 1`] = `
Object {
  "types": Array [
    Object {
      "container": null,
      "database": null,
      "fields": Array [
        Object {
          "container": "Dummies",
          "database": "Test",
          "fieldname": "dummies",
          "kind": "many-root",
          "ours": null,
          "pagination": true,
          "returnMany": true,
          "returnTypename": "Dummy",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "typename": "Query",
    },
    Object {
      "container": null,
      "database": null,
      "fields": Array [
        Object {
          "container": null,
          "database": null,
          "fieldname": "id",
          "kind": "embedded",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [
            "eq",
            "in",
          ],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "database": null,
          "fieldname": "text",
          "kind": "embedded",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "String",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "typename": "Dummy",
    },
  ],
}
`;
